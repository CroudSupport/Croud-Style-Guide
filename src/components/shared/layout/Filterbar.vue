<template>
    <div class="ui secondary basic segment">
        <div class="ui text menu">
            <div class="item">
                <div class="ui icon input">
                    <i class="search icon"></i>
                    <input  placeholder="Search..." v-model="_search"/>
                </div>
            </div>
            <div class="right item">
                <slot></slot>
            </div>
        </div>
    </div>
</template>

<script>
    /**
     * Filter bar to be used above datagrids and other searchable/filterable data sets
     *
     * @example ./croud-filterbar.md
     */
    export default {
        name: 'croud-filterbar',

        props: {
           /**
            * Search input, emitting update:search event compatible with prop.sync
            */
            search: {
                type: String,
                default: '',
            },
        },

        computed: {
            _search: {
                get() {
                    return this.search
                },
                set(val) {
                    this.$emit('update:search', val)
                },
            },
        },
    }
</script>

<style lang="scss" scoped>
    .ui.secondary.basic.segment {
        padding: 0.5em 0 0;
        margin: 0;

        .right.item {
            padding: 0;
            margin: 0;

            >.item {
                padding: 0 1em 1em;

                .segment {
                    padding: 0;
                    margin: 0;
                }
            }
        }
    }
</style>
