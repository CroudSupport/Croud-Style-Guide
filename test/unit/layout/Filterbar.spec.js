import Vue from 'vue'
import Filterbar from '../../../src/components/shared/layout/Filterbar'

const updateSearch = jest.fn()

const Constructor = Vue.extend({
    data() {
        return {
            showDefault: false,
            showAdvancedFilters: false,
        }
    },
    render(h) {
        return h(Filterbar, {
            on: {
                'update:search': updateSearch,
            },
        }, [
            this.showDefault ? h('div', {
                class: 'default-slot-test',
            }) : null,
            this.showAdvancedFilters ? h('div', {
                class: 'advanced-slot-test',
                slot: 'advanced-filters',
            }) : null,

        ])
    },
})

const vm = new Constructor().$mount()

describe('Filerbar', () => {
    it('should match the snapshot', () => {
        expect(vm.$el).toMatchSnapshot()
    })

    it('should emit new search to parent', () => {
        const newSearchTerm = 'netflix'
        vm.$children[0]._search = newSearchTerm
        expect(updateSearch).toBeCalledWith(newSearchTerm)
    })

    describe('slots', () => {
        beforeEach(() => {
            vm.showDefault = false
            vm.showAdvancedFilters = false
        })

        it('should show default props', (done) => {
            vm.showDefault = true
            vm.$nextTick(() => {
                expect(vm.$el.querySelectorAll('.default-slot-test').length).toBe(1)
                done()
            })
        })

        it('should hide advanced filters by default', (done) => {
            vm.showAdvancedFilters = true
            vm.$nextTick(() => {
                expect(vm.$children[0].collapse).toBe(true)
                expect(vm.$el.querySelector('button').textContent).toBe('Advanced')
                done()
            })
        })

        it('should show advanced filters when not collapsed', (done) => {
            vm.showAdvancedFilters = true
            vm.$children[0].collapse = false
            vm.$nextTick(() => {
                expect(vm.$el.querySelectorAll('.advanced-slot-test').length).toBe(1)
                done()
            })
        })
    })
})
