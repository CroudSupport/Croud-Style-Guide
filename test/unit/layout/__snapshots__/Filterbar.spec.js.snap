// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filerbar should match the snapshot 1`] = `
<div
  class="ui secondary basic segment"
>
  <div
    class="ui text menu"
  >
    <div
      class="left item"
    >
      <div
        class="ui icon input"
      >
        <i
          class="search icon"
        />
        <input
          placeholder="Search..."
        />
      </div>
    </div>
    <div
      class="right item"
    />
  </div>
</div>
`;
